Fix 19 frontend test failures (browser positioning and HTTP integration tests)

# Problem

19 frontend tests are currently failing. Analysis shows these are tests that require the HTTP dev server to be running with full database mode, but are failing even when endpoints are available.

## Failing Tests Breakdown

**Total:** 19 failures across 8 test files

### Category 1: Browser Positioning Tests (4 failures)

**Files:**
- `src/tests/browser/autocomplete-interaction.test.ts` (2 failures)
- `src/tests/browser/slash-dropdown-interaction.test.ts` (2 failures)

**Error Pattern:**
```
AssertionError: expected 0 to be greater than 0
```

**Tests:**
1. "verifies getBoundingClientRect returns real dimensions" (autocomplete)
2. "can calculate dropdown position relative to cursor" (autocomplete)
3. "verifies getBoundingClientRect for dropdown positioning" (slash)
4. "can calculate cursor position for dropdown placement" (slash)

**Root Cause:** `getBoundingClientRect()` returns 0 dimensions in browser test environment.

---

### Category 2: Integration Tests Requiring HTTP Server (12 failures)

**Files:**
- `src/tests/services/database-persistence.test.ts` (suite-level failure)
- `src/tests/services/event-emission.test.ts` (suite-level failure)
- `src/tests/services/node-ordering.test.ts` (suite-level failure)
- `src/tests/integration/edge-cases.test.ts` (3 tests)
- `src/tests/integration/integration-scenarios.test.ts` (3 tests)
- `src/tests/integration/regression-prevention.test.ts` (6 tests)

**Common Pattern:** Tests attempting to use HTTP endpoints but failing with various errors.

**Examples:**
- "should cleanup resources on timeout"
- "should handle mention creation with non-existent nodes"
- "should create container node and establish mention relationship"
- "should maintain sibling order for container nodes"
- "should handle concurrent container node creation"

**Root Cause:** These tests expect HTTP dev server but are running in default in-memory mode, or endpoints are returning errors.

---

### Category 3: Suite-Level Failures (3 test suites)

**Files:**
- `src/tests/services/database-persistence.test.ts`
- `src/tests/services/event-emission.test.ts`
- `src/tests/services/node-ordering.test.ts`

**Issue:** Entire test suites failing (not individual tests).

**Root Cause:** Likely setup/teardown issues or missing database initialization.

---

## Investigation Needed

1. **Check if tests should be running in in-memory mode or database mode**
   - If database mode only: Should use `it.skipIf(!shouldUseDatabase())`
   - If in-memory compatible: Need to fix the test implementation

2. **Browser positioning tests**
   - Verify Playwright browser is rendering properly
   - Check if elements are mounted before getBoundingClientRect calls
   - May need `waitFor` or similar async coordination

3. **Suite-level failures**
   - Review beforeAll/afterAll hooks
   - Check database initialization in test setup
   - Verify cleanup between tests

---

## Reproduction

```bash
# Default in-memory mode (19 failures)
bun run test

# Database mode (need to verify behavior)
bun run test:db

# Browser tests specifically
bun run test:browser
```

---

## Acceptance Criteria

- [ ] All 19 failing tests pass
- [ ] Browser positioning tests return non-zero dimensions
- [ ] Integration tests properly skip in in-memory mode OR work in in-memory mode
- [ ] Suite-level test setup/teardown fixed
- [ ] Root cause documented for each failure category
- [ ] Tests pass: `bun run test`

---

## Related Issues

- #282 (Browser Mode infrastructure - complete)
- #406 (closed - addressed backend, but frontend failures remain)

---

## Priority

🔴 **HIGH** - 19 failing tests reduce confidence in frontend codebase.

---

## Technical Specifications

### Reference Files

**Browser Tests:**
- `packages/desktop-app/src/tests/browser/autocomplete-interaction.test.ts`
- `packages/desktop-app/src/tests/browser/slash-dropdown-interaction.test.ts`

**Integration Tests:**
- `packages/desktop-app/src/tests/integration/edge-cases.test.ts`
- `packages/desktop-app/src/tests/integration/integration-scenarios.test.ts`
- `packages/desktop-app/src/tests/integration/regression-prevention.test.ts`

**Service Tests:**
- `packages/desktop-app/src/tests/services/database-persistence.test.ts`
- `packages/desktop-app/src/tests/services/event-emission.test.ts`
- `packages/desktop-app/src/tests/services/node-ordering.test.ts`

--- a/TEST-STATE-ANALYSIS.md
+++ b/TEST-STATE-ANALYSIS.md
@@ -1,0 +1,260 @@
+# NodeSpace Test State Analysis - Issue #409
+**Date:** 2025-11-04
+**Objective:** Achieve 100% test pass rate across all test suites
+
+## Executive Summary
+
+### Current State
+- **Frontend Tests:** 19 failures out of 1611 tests (98.8% pass rate)
+- **Browser Tests:** 77/77 passing (100% ✅)
+- **Rust Backend Tests:** 6 failures out of 391 tests (98.5% pass rate)
+- **Total Failures:** 25 tests failing across all suites
+
+### Goal
+Achieve 100% test pass rate - all frontend, browser, and backend tests passing.
+
+---
+
+## Detailed Breakdown
+
+### 1. Frontend Tests (19 Failures)
+
+**Status:** 19 failed | 1559 passed | 33 skipped out of 1611 total tests
+
+**Root Cause:** All 19 failures are HTTP integration tests attempting to connect to `http://localhost:3001` which is not running.
+
+#### Error Pattern
+```
+NodeOperationError: Failed to execute "fetch()" on "Window" with URL "http://localhost:3001/api/nodes"
+AggregateError: ECONNREFUSED ::1:3001 / 127.0.0.1:3001
+```
+
+#### Failing Test Files
+
+1. **`src/tests/integration/regression-prevention.test.ts`** (12 failures)
+   - Hierarchy integrity tests (2 tests):
+     - "should maintain sibling order for container nodes"
+     - "should preserve parent-child relationships during mention operations"
+   - Mention cascade deletion tests (2 tests):
+     - "should remove mention relationships when mentioned node is deleted"
+     - "should remove mention relationships when mentioning node is deleted"
+   - Concurrent operation safety (2 tests):
+     - "should handle concurrent container node creation"
+     - "should handle concurrent mention creation to same node"
+   - Content change detection tests (6 tests):
+     - "should mark topic as stale when content changes"
+     - "should trigger re-embedding on topic close after edit"
+     - "should trigger re-embedding on idle timeout"
+     - And 3 others with similar patterns
+
+2. **`src/tests/integration/integration-scenarios.test.ts`** (3 failures)
+   - Tests requiring HTTP endpoints for complex workflows
+
+3. **`src/tests/integration/edge-cases.test.ts`** (3 failures)
+   - Edge case tests requiring database persistence
+
+4. **`src/tests/services/database-persistence.test.ts`** (suite-level failure)
+   - Entire suite failing due to HTTP connection issues
+
+5. **`src/tests/services/event-emission.test.ts`** (suite-level failure)
+   - Event emission tests requiring HTTP backend
+
+6. **`src/tests/services/node-ordering.test.ts`** (suite-level failure)
+   - Node ordering tests requiring HTTP backend
+
+#### Solution Approach
+
+**Option A: Skip in Default Mode** (Recommended)
+```typescript
+import { shouldUseDatabase } from '$tests/test-utils';
+
+describe.skipIf(!shouldUseDatabase())('Integration Tests Requiring HTTP', () => {
+  // Tests here will only run in database mode (TEST_USE_DATABASE=true)
+});
+```
+
+**Option B: Make In-Memory Compatible**
+- Refactor tests to work with in-memory adapter
+- May require significant test rewrites
+- Not all tests can work in-memory (e.g., concurrent operations)
+
+**Recommendation:** Use Option A - these are integration tests that specifically test HTTP/database behavior and should only run in database mode.
+
+---
+
+### 2. Browser Tests (✅ All Passing)
+
+**Status:** 77/77 tests passing (100% pass rate)
+
+**Test Files:**
+- `src/tests/browser/autocomplete-interaction.test.ts` (17 tests) ✅
+- `src/tests/browser/slash-dropdown-interaction.test.ts` (21 tests) ✅
+- `src/tests/browser/enter-key-focus.test.ts` (15 tests) ✅
+- `src/tests/browser/backspace-focus.test.ts` (18 tests) ✅
+- `src/tests/browser/focus-management.test.ts` (6 tests) ✅
+
+**Note:** Issue #409 originally mentioned 4 browser positioning test failures, but these are now passing. The fixes from previous issues (#407, #282) have resolved these.
+
+---
+
+### 3. Rust Backend Tests (6 Failures)
+
+**Status:** 6 failed | 379 passed | 6 ignored out of 391 total tests
+
+**Test Suite Results:**
+- `nodespace-app` lib tests: 26/26 passing ✅
+- `nodespace-app` main tests: 0 tests (no failures) ✅
+- `nodespace-core` lib tests: 6 failures ❌
+
+#### Failing Tests
+
+1. **`mcp::handlers::markdown::markdown_test::tests::test_content_preservation`**
+   - Markdown content preservation failing
+   - Location: `packages/core/src/mcp/handlers/markdown/`
+
+2. **`mcp::handlers::markdown::markdown_test::tests::test_mixed_content`**
+   - Mixed content handling in markdown import
+   - Location: `packages/core/src/mcp/handlers/markdown/`
+
+3. **`operations::sibling_queue::tests::test_reorder_with_retry_exponential_backoff`**
+   - Sibling reordering retry logic failing
+   - Location: `packages/core/src/operations/sibling_queue.rs`
+
+4. **`operations::tests::test_concurrent_move_version_conflict`**
+   - Concurrent node move with version conflicts
+   - Location: `packages/core/src/operations/mod.rs`
+
+5. **`operations::tests::test_version_conflict_includes_current_state`**
+   - Version conflict state reporting
+   - Location: `packages/core/src/operations/mod.rs`
+
+6. **`mcp::handlers::nodes::nodes_test::integration_tests::test_occ_performance_overhead`**
+   - Optimistic concurrency control (OCC) performance test
+   - Location: `packages/core/src/mcp/handlers/nodes/`
+
+#### Solution Approach
+
+**Investigation Needed:**
+1. Run individual tests with `RUST_BACKTRACE=full` to see detailed error messages
+2. Check for recent changes to:
+   - Markdown import/export logic
+   - Sibling ordering/retry mechanisms
+   - Version conflict handling
+   - OCC performance benchmarks
+
+**Likely Causes:**
+- Schema changes affecting markdown serialization
+- Timing issues in retry logic tests
+- Version conflict edge cases not handled
+- Performance thresholds too strict for CI environment
+
+---
+
+## Reproduction Commands
+
+### Frontend Tests
+```bash
+# In-memory mode (19 failures expected)
+bun run test
+
+# Database mode (need to verify if failures persist)
+bun run test:db
+
+# Browser tests (all passing)
+bun run test:browser
+
+# All tests
+bun run test:all
+```
+
+### Backend Tests
+```bash
+# Run all Rust tests
+cargo test
+
+# Run specific failing test with backtrace
+RUST_BACKTRACE=full cargo test test_content_preservation -- --nocapture
+
+# Run all tests with output
+cargo test -- --nocapture
+```
+
+---
+
+## Action Plan
+
+### Phase 1: Frontend Fixes (Priority: HIGH)
+1. ✅ Identify all 19 failing frontend tests
+2. ⏳ Add `skipIf(!shouldUseDatabase())` to integration tests
+3. ⏳ Verify tests pass in database mode
+4. ⏳ Document which tests require HTTP backend
+
+### Phase 2: Rust Backend Fixes (Priority: HIGH)
+1. ⏳ Run failing tests individually with full backtrace
+2. ⏳ Analyze error messages and root causes
+3. ⏳ Fix markdown content preservation issues (2 tests)
+4. ⏳ Fix sibling queue retry logic (1 test)
+5. ⏳ Fix version conflict handling (2 tests)
+6. ⏳ Fix or adjust OCC performance test (1 test)
+
+### Phase 3: Verification (Priority: HIGH)
+1. ⏳ Run full test suite: `bun run test` (expect 0 failures)
+2. ⏳ Run browser tests: `bun run test:browser` (expect 77/77)
+3. ⏳ Run database tests: `bun run test:db` (expect all passing)
+4. ⏳ Run Rust tests: `cargo test` (expect 391/391)
+5. ⏳ Verify total: 0 failures across all test suites
+
+---
+
+## Success Criteria
+
+- [ ] All 1611 frontend tests pass (currently 1559/1611)
+- [ ] All 77 browser tests continue passing (currently 77/77 ✅)
+- [ ] All 391 Rust backend tests pass (currently 379/391)
+- [ ] Integration tests properly skip in in-memory mode
+- [ ] Integration tests pass in database mode
+- [ ] Root cause documented for each failure
+- [ ] Tests remain stable in CI/CD
+
+---
+
+## Technical Context
+
+### Test Modes
+NodeSpace uses a hybrid testing strategy:
+
+1. **In-Memory Mode (Default - Fast)**
+   - `bun run test` - 100x faster, perfect for TDD
+   - Tests run with mocked HTTP adapter
+   - No database persistence
+   - Ideal for unit tests and logic validation
+
+2. **Database Mode (Full Integration)**
+   - `bun run test:db` - Full SQLite persistence
+   - Requires HTTP dev server on port 3001
+   - Tests real backend integration
+   - Required for integration tests
+
+3. **Browser Mode (Real DOM)**
+   - `bun run test:browser` - Real Chromium via Playwright
+   - Required for focus/blur events
+   - Real browser DOM API behavior
+   - Currently all 77 tests passing ✅
+
+### Related Issues
+- #282 - Browser Mode infrastructure (complete)
+- #406 - Backend test failures (closed - but frontend issues remain)
+- #407 - Fixed 23 test failures (complete)
+- #408 - Investigated SIGABRT crash in Rust tests (complete)
+- #405 - Fixed test isolation issues (complete)
+
+---
+
+## Next Steps
+
+1. **Immediate:** Fix frontend integration tests (add `skipIf` guards)
+2. **Short-term:** Debug and fix 6 Rust backend test failures
+3. **Validation:** Run full test suite to verify 100% pass rate
+4. **Documentation:** Update testing guide with findings
+
+**Estimated Time:** 2-4 hours for frontend fixes + 2-3 hours for Rust fixes = 4-7 hours total

--- a/packages/desktop-app/src/tests/integration/edge-cases.test.ts
+++ b/packages/desktop-app/src/tests/integration/edge-cases.test.ts
@@ -229,42 +229,46 @@ describe.sequential('Section 10: Edge Cases & Error Handling', () => {
       35000
     ); // Allow 35s for timeout test + cleanup
 
-    it('should cleanup resources on timeout', async () => {
-      try {
-        // Create multiple topics
-        const topic1Data = TestNodeBuilder.text('Topic 1').build();
-        const topic1Id = await backend.createNode(topic1Data);
-
-        const topic2Data = TestNodeBuilder.text('Topic 2').build();
-        const topic2Id = await backend.createNode(topic2Data);
-
-        await waitForDatabaseWrites();
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-
-        // Batch operation (placeholder will fail fast)
+    it.skipIf(!shouldUseDatabase())(
+      'should cleanup resources on timeout',
+      async () => {
         try {
-          await backend.batchGenerateEmbeddings([topic1Id, topic2Id]);
-        } catch (error) {
-          // Verify error is thrown
-          expect(error).toBeTruthy();
-        }
+          // Create multiple topics
+          const topic1Data = TestNodeBuilder.text('Topic 1').build();
+          const topic1Id = await backend.createNode(topic1Data);
 
-        // Verify nodes still exist (cleanup should not delete nodes)
-        const topic1 = await backend.getNode(topic1Id);
-        const topic2 = await backend.getNode(topic2Id);
-        expect(topic1).toBeTruthy();
-        expect(topic2).toBeTruthy();
-      } catch (error) {
-        // If getNode returns 500 error, skip this test
-        const errorMessage = error instanceof Error ? error.message : String(error);
-        if (errorMessage.includes('500')) {
-          console.log('[Test] Node retrieval endpoint error - test skipped');
-          expect(error).toBeTruthy();
-        } else {
-          throw error;
+          const topic2Data = TestNodeBuilder.text('Topic 2').build();
+          const topic2Id = await backend.createNode(topic2Data);
+
+          await waitForDatabaseWrites();
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+
+          // Batch operation (placeholder will fail fast)
+          try {
+            await backend.batchGenerateEmbeddings([topic1Id, topic2Id]);
+          } catch (error) {
+            // Verify error is thrown
+            expect(error).toBeTruthy();
+          }
+
+          // Verify nodes still exist (cleanup should not delete nodes)
+          const topic1 = await backend.getNode(topic1Id);
+          const topic2 = await backend.getNode(topic2Id);
+          expect(topic1).toBeTruthy();
+          expect(topic2).toBeTruthy();
+        } catch (error) {
+          // If getNode returns 500 error, skip this test
+          const errorMessage = error instanceof Error ? error.message : String(error);
+          if (errorMessage.includes('500')) {
+            console.log('[Test] Node retrieval endpoint error - test skipped');
+            expect(error).toBeTruthy();
+          } else {
+            throw error;
+          }
         }
-      }
-    }, 35000);
+      },
+      35000
+    );
   });
 
   describe('Error propagation', () => {
@@ -337,7 +341,7 @@ describe.sequential('Section 10: Edge Cases & Error Handling', () => {
     );
   });
 
-  describe('Mention operation edge cases', () => {
+  describe.skipIf(!shouldUseDatabase())('Mention operation edge cases', () => {
     it('should handle mention creation with non-existent nodes', async () => {
       try {
         // Test with non-existent mentioning node

--- a/packages/desktop-app/src/tests/integration/integration-scenarios.test.ts
+++ b/packages/desktop-app/src/tests/integration/integration-scenarios.test.ts
@@ -49,299 +49,311 @@ describe.sequential('Section 11: Integration Scenarios', () => {
     sharedNodeStore.clearTestErrors();
   });
 
-  it('should create container node and establish mention relationship', async () => {
-    try {
-      // Step 1: Create a daily note (the mentioning node)
-      const dailyNoteData = TestNodeBuilder.text('Daily Note 2025-01-15')
-        .withType('date')
-        .withId('2025-01-15')
-        .build();
-      const dailyNoteId = await backend.createNode(dailyNoteData);
-
-      await waitForDatabaseWrites();
-      if (shouldUseDatabase()) {
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-      }
-
-      // Step 2: Create a container node with mention
-      const containerInput = {
-        content: 'Project: NodeSpace Testing',
-        nodeType: 'text',
-        properties: { status: 'active', priority: 'high' },
-        mentionedBy: dailyNoteId
-      };
-      const containerId = await backend.createContainerNode(containerInput);
-
-      // Step 3: Verify container was created correctly
-      const container = await backend.getNode(containerId);
-      expect(container).toBeTruthy();
-      expect(container?.content).toBe('Project: NodeSpace Testing');
-      expect(container?.parentId).toBeNull(); // Container nodes are root nodes
-      expect(container?.properties).toEqual({ status: 'active', priority: 'high' });
-
-      // Step 4: Create explicit mention relationship (in addition to mentionedBy)
-      await backend.createNodeMention(dailyNoteId, containerId);
-
-      // Step 5: Create child nodes under the container
-      const task1Data = TestNodeBuilder.text('Implement Phase 3 tests')
-        .withParent(containerId)
-        .withContainer(containerId)
-        .build();
-      const task1Id = await backend.createNode(task1Data);
-
-      await waitForDatabaseWrites();
-      if (shouldUseDatabase()) {
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-      }
-
-      const task2Data = TestNodeBuilder.text('Review and merge PR')
-        .withParent(containerId)
-        .withContainer(containerId)
-        .withBeforeSibling(task1Id)
-        .build();
-      const task2Id = await backend.createNode(task2Data);
-
-      await waitForDatabaseWrites();
-      if (shouldUseDatabase()) {
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-      }
-
-      // Step 6: Verify hierarchy and relationships
-      const children = await backend.queryNodes({ parentId: containerId });
-      expect(children.length).toBe(2);
-
-      const task1 = await backend.getNode(task1Id);
-      const task2 = await backend.getNode(task2Id);
-      expect(task1?.containerNodeId).toBe(containerId);
-      expect(task2?.containerNodeId).toBe(containerId);
-      expect(task1?.parentId).toBe(containerId);
-      expect(task2?.parentId).toBe(containerId);
-
-      // Step 7: Verify daily note still exists
-      const dailyNote = await backend.getNode(dailyNoteId);
-      expect(dailyNote).toBeTruthy();
-    } catch (error) {
-      // If container or mention endpoints are not yet active, skip this test
-      // Expected: 405 Method Not Allowed until endpoints are properly registered
-      if (skipIfEndpointUnavailable(error, 'Container/mention endpoints')) {
-        expect(error).toBeTruthy();
-        return;
-      }
-      throw error;
-    }
-  }, 10000);
-
-  it('should handle embedding lifecycle: create → update → search', async () => {
-    try {
-      // Step 1: Create topic nodes with initial content
-      const topic1Data = TestNodeBuilder.text('Machine Learning Fundamentals')
-        .withProperties({ category: 'AI', difficulty: 'beginner' })
-        .build();
-      const topic1Id = await backend.createNode(topic1Data);
-
-      const topic2Data = TestNodeBuilder.text('Deep Learning and Neural Networks')
-        .withProperties({ category: 'AI', difficulty: 'advanced' })
-        .build();
-      const topic2Id = await backend.createNode(topic2Data);
-
-      const topic3Data = TestNodeBuilder.text('Natural Language Processing')
-        .withProperties({ category: 'AI', difficulty: 'intermediate' })
-        .build();
-      const topic3Id = await backend.createNode(topic3Data);
-
-      await waitForDatabaseWrites();
-      if (shouldUseDatabase()) {
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-      }
-
-      // Step 2: Generate embeddings for all topics
-      // Note: Placeholder implementation will return NOT_IMPLEMENTED
+  it.skipIf(!shouldUseDatabase())(
+    'should create container node and establish mention relationship',
+    async () => {
       try {
-        await backend.generateContainerEmbedding(topic1Id);
-        await backend.generateContainerEmbedding(topic2Id);
-        await backend.generateContainerEmbedding(topic3Id);
+        // Step 1: Create a daily note (the mentioning node)
+        const dailyNoteData = TestNodeBuilder.text('Daily Note 2025-01-15')
+          .withType('date')
+          .withId('2025-01-15')
+          .build();
+        const dailyNoteId = await backend.createNode(dailyNoteData);
+
+        await waitForDatabaseWrites();
+        if (shouldUseDatabase()) {
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+        }
+
+        // Step 2: Create a container node with mention
+        const containerInput = {
+          content: 'Project: NodeSpace Testing',
+          nodeType: 'text',
+          properties: { status: 'active', priority: 'high' },
+          mentionedBy: dailyNoteId
+        };
+        const containerId = await backend.createContainerNode(containerInput);
+
+        // Step 3: Verify container was created correctly
+        const container = await backend.getNode(containerId);
+        expect(container).toBeTruthy();
+        expect(container?.content).toBe('Project: NodeSpace Testing');
+        expect(container?.parentId).toBeNull(); // Container nodes are root nodes
+        expect(container?.properties).toEqual({ status: 'active', priority: 'high' });
+
+        // Step 4: Create explicit mention relationship (in addition to mentionedBy)
+        await backend.createNodeMention(dailyNoteId, containerId);
+
+        // Step 5: Create child nodes under the container
+        const task1Data = TestNodeBuilder.text('Implement Phase 3 tests')
+          .withParent(containerId)
+          .withContainer(containerId)
+          .build();
+        const task1Id = await backend.createNode(task1Data);
+
+        await waitForDatabaseWrites();
+        if (shouldUseDatabase()) {
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+        }
+
+        const task2Data = TestNodeBuilder.text('Review and merge PR')
+          .withParent(containerId)
+          .withContainer(containerId)
+          .withBeforeSibling(task1Id)
+          .build();
+        const task2Id = await backend.createNode(task2Data);
+
+        await waitForDatabaseWrites();
+        if (shouldUseDatabase()) {
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+        }
+
+        // Step 6: Verify hierarchy and relationships
+        const children = await backend.queryNodes({ parentId: containerId });
+        expect(children.length).toBe(2);
+
+        const task1 = await backend.getNode(task1Id);
+        const task2 = await backend.getNode(task2Id);
+        expect(task1?.containerNodeId).toBe(containerId);
+        expect(task2?.containerNodeId).toBe(containerId);
+        expect(task1?.parentId).toBe(containerId);
+        expect(task2?.parentId).toBe(containerId);
+
+        // Step 7: Verify daily note still exists
+        const dailyNote = await backend.getNode(dailyNoteId);
+        expect(dailyNote).toBeTruthy();
       } catch (error) {
-        // Expected: NOT_IMPLEMENTED until TopicEmbeddingService is integrated
-        expect(error).toBeTruthy();
+        // If container or mention endpoints are not yet active, skip this test
+        // Expected: 405 Method Not Allowed until endpoints are properly registered
+        if (skipIfEndpointUnavailable(error, 'Container/mention endpoints')) {
+          expect(error).toBeTruthy();
+          return;
+        }
+        throw error;
       }
+    },
+    10000
+  );
 
-      // Step 3: Update content of topic1 (should mark as stale)
-      await backend.updateNode(topic1Id, 1, {
-        content: 'Introduction to Machine Learning and AI'
-      });
-
-      await waitForDatabaseWrites();
-      if (shouldUseDatabase()) {
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-      }
-
-      // Verify content was updated
-      const updatedTopic1 = await backend.getNode(topic1Id);
-      expect(updatedTopic1?.content).toBe('Introduction to Machine Learning and AI');
-
-      // Step 4: Trigger re-embedding (smart trigger: on topic closed)
+  it.skipIf(!shouldUseDatabase())(
+    'should handle embedding lifecycle: create → update → search',
+    async () => {
       try {
-        await backend.onContainerClosed(topic1Id);
-      } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
-      }
+        // Step 1: Create topic nodes with initial content
+        const topic1Data = TestNodeBuilder.text('Machine Learning Fundamentals')
+          .withProperties({ category: 'AI', difficulty: 'beginner' })
+          .build();
+        const topic1Id = await backend.createNode(topic1Data);
 
-      // Step 5: Search for related topics
-      try {
-        const searchResults = await backend.searchContainers({
-          query: 'machine learning',
-          threshold: 0.7,
-          limit: 10
+        const topic2Data = TestNodeBuilder.text('Deep Learning and Neural Networks')
+          .withProperties({ category: 'AI', difficulty: 'advanced' })
+          .build();
+        const topic2Id = await backend.createNode(topic2Data);
+
+        const topic3Data = TestNodeBuilder.text('Natural Language Processing')
+          .withProperties({ category: 'AI', difficulty: 'intermediate' })
+          .build();
+        const topic3Id = await backend.createNode(topic3Data);
+
+        await waitForDatabaseWrites();
+        if (shouldUseDatabase()) {
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+        }
+
+        // Step 2: Generate embeddings for all topics
+        // Note: Placeholder implementation will return NOT_IMPLEMENTED
+        try {
+          await backend.generateContainerEmbedding(topic1Id);
+          await backend.generateContainerEmbedding(topic2Id);
+          await backend.generateContainerEmbedding(topic3Id);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED until TopicEmbeddingService is integrated
+          expect(error).toBeTruthy();
+        }
+
+        // Step 3: Update content of topic1 (should mark as stale)
+        await backend.updateNode(topic1Id, 1, {
+          content: 'Introduction to Machine Learning and AI'
         });
 
-        // In real implementation, should return relevant topics
-        expect(Array.isArray(searchResults)).toBe(true);
+        await waitForDatabaseWrites();
+        if (shouldUseDatabase()) {
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+        }
 
-        // Results should be sorted by similarity (highest first)
-        // Each result should have required node fields
-        for (const node of searchResults) {
-          expect(node).toHaveProperty('id');
-          expect(node).toHaveProperty('content');
-          expect(node).toHaveProperty('nodeType');
+        // Verify content was updated
+        const updatedTopic1 = await backend.getNode(topic1Id);
+        expect(updatedTopic1?.content).toBe('Introduction to Machine Learning and AI');
+
+        // Step 4: Trigger re-embedding (smart trigger: on topic closed)
+        try {
+          await backend.onContainerClosed(topic1Id);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
+
+        // Step 5: Search for related topics
+        try {
+          const searchResults = await backend.searchContainers({
+            query: 'machine learning',
+            threshold: 0.7,
+            limit: 10
+          });
+
+          // In real implementation, should return relevant topics
+          expect(Array.isArray(searchResults)).toBe(true);
+
+          // Results should be sorted by similarity (highest first)
+          // Each result should have required node fields
+          for (const node of searchResults) {
+            expect(node).toHaveProperty('id');
+            expect(node).toHaveProperty('content');
+            expect(node).toHaveProperty('nodeType');
+          }
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
+
+        // Step 6: Check stale count
+        try {
+          const staleCount = await backend.getStaleContainerCount();
+          expect(typeof staleCount).toBe('number');
+          expect(staleCount).toBeGreaterThanOrEqual(0);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
+
+        // Step 7: Sync all embeddings
+        try {
+          const syncedCount = await backend.syncEmbeddings();
+          expect(typeof syncedCount).toBe('number');
+          expect(syncedCount).toBeGreaterThanOrEqual(0);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
         }
       } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
+        // If updateNode returns 500 error, skip this test
+        const errorMessage = error instanceof Error ? error.message : String(error);
+        if (errorMessage.includes('500')) {
+          console.log('[Test] Node update endpoint error - test skipped');
+          expect(error).toBeTruthy();
+        } else {
+          throw error;
+        }
       }
+    },
+    15000
+  );
 
-      // Step 6: Check stale count
+  it.skipIf(!shouldUseDatabase())(
+    'should coordinate stale tracking with smart triggers',
+    async () => {
       try {
-        const staleCount = await backend.getStaleContainerCount();
-        expect(typeof staleCount).toBe('number');
-        expect(staleCount).toBeGreaterThanOrEqual(0);
+        // Step 1: Create a topic node
+        const topicData = TestNodeBuilder.text('TypeScript Best Practices')
+          .withProperties({ language: 'TypeScript', level: 'advanced' })
+          .build();
+        const topicId = await backend.createNode(topicData);
+
+        await waitForDatabaseWrites();
+        if (shouldUseDatabase()) {
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+        }
+
+        // Step 2: Generate initial embedding
+        try {
+          await backend.generateContainerEmbedding(topicId);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
+
+        // Step 3: Update content multiple times (simulate editing)
+        await backend.updateNode(topicId, 1, {
+          content: 'TypeScript Best Practices - Updated'
+        });
+
+        await backend.updateNode(topicId, 2, {
+          content: 'TypeScript Best Practices - Final Version'
+        });
+
+        await waitForDatabaseWrites();
+        if (shouldUseDatabase()) {
+          expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
+        }
+
+        // Step 4: Check stale count (should be at least 1)
+        try {
+          const staleCount = await backend.getStaleContainerCount();
+          expect(typeof staleCount).toBe('number');
+
+          // In real implementation, should be >= 1 after content changes
+          console.log(`Stale topics count: ${staleCount}`);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
+
+        // Step 5: Trigger re-embedding via idle timeout
+        try {
+          const wasTriggered = await backend.onContainerIdle(topicId);
+          expect(typeof wasTriggered).toBe('boolean');
+
+          // In real implementation:
+          // - Returns true if re-embedding was triggered (topic was stale)
+          // - Returns false if topic was not stale
+          console.log(`Re-embedding triggered: ${wasTriggered}`);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
+
+        // Step 6: Verify topic still exists and has final content
+        const finalTopic = await backend.getNode(topicId);
+        if (!finalTopic) {
+          // If node doesn't exist, updateNode might have failed with 500 error
+          // This is acceptable during dev server rebuild
+          console.log('[Test] Node not found after update - likely endpoint error');
+          expect(finalTopic).toBeNull();
+          return; // Skip remaining assertions
+        }
+        expect(finalTopic).toBeTruthy();
+        expect(finalTopic?.content).toBe('TypeScript Best Practices - Final Version');
+
+        // Step 7: Trigger re-embedding via topic closed
+        try {
+          await backend.onContainerClosed(topicId);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
+
+        // Step 8: Check stale count again (should be 0 after re-embedding)
+        try {
+          const staleCountAfter = await backend.getStaleContainerCount();
+          expect(typeof staleCountAfter).toBe('number');
+
+          // In real implementation, should be 0 after successful re-embedding
+          console.log(`Stale topics count after sync: ${staleCountAfter}`);
+        } catch (error) {
+          // Expected: NOT_IMPLEMENTED
+          expect(error).toBeTruthy();
+        }
       } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
+        // If updateNode returns 500 error, skip this test
+        const errorMessage = error instanceof Error ? error.message : String(error);
+        if (errorMessage.includes('500')) {
+          console.log('[Test] Node update endpoint error - test skipped');
+          expect(error).toBeTruthy();
+        } else {
+          throw error;
+        }
       }
-
-      // Step 7: Sync all embeddings
-      try {
-        const syncedCount = await backend.syncEmbeddings();
-        expect(typeof syncedCount).toBe('number');
-        expect(syncedCount).toBeGreaterThanOrEqual(0);
-      } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
-      }
-    } catch (error) {
-      // If updateNode returns 500 error, skip this test
-      const errorMessage = error instanceof Error ? error.message : String(error);
-      if (errorMessage.includes('500')) {
-        console.log('[Test] Node update endpoint error - test skipped');
-        expect(error).toBeTruthy();
-      } else {
-        throw error;
-      }
-    }
-  }, 15000);
-
-  it('should coordinate stale tracking with smart triggers', async () => {
-    try {
-      // Step 1: Create a topic node
-      const topicData = TestNodeBuilder.text('TypeScript Best Practices')
-        .withProperties({ language: 'TypeScript', level: 'advanced' })
-        .build();
-      const topicId = await backend.createNode(topicData);
-
-      await waitForDatabaseWrites();
-      if (shouldUseDatabase()) {
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-      }
-
-      // Step 2: Generate initial embedding
-      try {
-        await backend.generateContainerEmbedding(topicId);
-      } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
-      }
-
-      // Step 3: Update content multiple times (simulate editing)
-      await backend.updateNode(topicId, 1, {
-        content: 'TypeScript Best Practices - Updated'
-      });
-
-      await backend.updateNode(topicId, 2, {
-        content: 'TypeScript Best Practices - Final Version'
-      });
-
-      await waitForDatabaseWrites();
-      if (shouldUseDatabase()) {
-        expect(sharedNodeStore.getTestErrors()).toHaveLength(0);
-      }
-
-      // Step 4: Check stale count (should be at least 1)
-      try {
-        const staleCount = await backend.getStaleContainerCount();
-        expect(typeof staleCount).toBe('number');
-
-        // In real implementation, should be >= 1 after content changes
-        console.log(`Stale topics count: ${staleCount}`);
-      } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
-      }
-
-      // Step 5: Trigger re-embedding via idle timeout
-      try {
-        const wasTriggered = await backend.onContainerIdle(topicId);
-        expect(typeof wasTriggered).toBe('boolean');
-
-        // In real implementation:
-        // - Returns true if re-embedding was triggered (topic was stale)
-        // - Returns false if topic was not stale
-        console.log(`Re-embedding triggered: ${wasTriggered}`);
-      } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
-      }
-
-      // Step 6: Verify topic still exists and has final content
-      const finalTopic = await backend.getNode(topicId);
-      if (!finalTopic) {
-        // If node doesn't exist, updateNode might have failed with 500 error
-        // This is acceptable during dev server rebuild
-        console.log('[Test] Node not found after update - likely endpoint error');
-        expect(finalTopic).toBeNull();
-        return; // Skip remaining assertions
-      }
-      expect(finalTopic).toBeTruthy();
-      expect(finalTopic?.content).toBe('TypeScript Best Practices - Final Version');
-
-      // Step 7: Trigger re-embedding via topic closed
-      try {
-        await backend.onContainerClosed(topicId);
-      } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
-      }
-
-      // Step 8: Check stale count again (should be 0 after re-embedding)
-      try {
-        const staleCountAfter = await backend.getStaleContainerCount();
-        expect(typeof staleCountAfter).toBe('number');
-
-        // In real implementation, should be 0 after successful re-embedding
-        console.log(`Stale topics count after sync: ${staleCountAfter}`);
-      } catch (error) {
-        // Expected: NOT_IMPLEMENTED
-        expect(error).toBeTruthy();
-      }
-    } catch (error) {
-      // If updateNode returns 500 error, skip this test
-      const errorMessage = error instanceof Error ? error.message : String(error);
-      if (errorMessage.includes('500')) {
-        console.log('[Test] Node update endpoint error - test skipped');
-        expect(error).toBeTruthy();
-      } else {
-        throw error;
-      }
-    }
-  }, 15000);
+    },
+    15000
+  );
 });

--- a/packages/desktop-app/src/tests/integration/regression-prevention.test.ts
+++ b/packages/desktop-app/src/tests/integration/regression-prevention.test.ts
@@ -91,7 +91,7 @@ describe.sequential('Section 12: Regression Prevention', () => {
     return sorted;
   }
 
-  describe('Hierarchy integrity (Regression #185, #176)', () => {
+  describe.skipIf(!shouldUseDatabase())('Hierarchy integrity (Regression #185, #176)', () => {
     it('should maintain sibling order for container nodes', async () => {
       try {
         // Issue #185: Lost track of nodes, sibling chain corruption
@@ -266,7 +266,7 @@ describe.sequential('Section 12: Regression Prevention', () => {
     }, 10000);
   });
 
-  describe('Mention cascade deletion (Regression #184)', () => {
+  describe.skipIf(!shouldUseDatabase())('Mention cascade deletion (Regression #184)', () => {
     it('should remove mention relationships when mentioned node is deleted', async () => {
       try {
         // Issue #184: Orphaned mentions after node deletion
@@ -383,7 +383,7 @@ describe.sequential('Section 12: Regression Prevention', () => {
     }, 10000);
   });
 
-  describe('Concurrent operation safety (Regression #190)', () => {
+  describe.skipIf(!shouldUseDatabase())('Concurrent operation safety (Regression #190)', () => {
     it('should handle concurrent container node creation', async () => {
       try {
         // Issue #190: Race conditions in hierarchy operations
@@ -486,7 +486,7 @@ describe.sequential('Section 12: Regression Prevention', () => {
     }, 10000);
   });
 
-  describe('Content change detection (Regression #187)', () => {
+  describe.skipIf(!shouldUseDatabase())('Content change detection (Regression #187)', () => {
     it('should mark topic as stale when content changes', async () => {
       try {
         // Issue #187: Embeddings not updated after content changes

--- a/packages/desktop-app/src/tests/services/database-persistence.test.ts
+++ b/packages/desktop-app/src/tests/services/database-persistence.test.ts
@@ -33,6 +33,7 @@ import {
   cleanDatabase,
   waitForDatabaseWrites
 } from '../utils/test-database';
+import { shouldUseDatabase } from '../utils/should-use-database';
 import { TestNodeBuilder } from '../utils/test-node-builder';
 import { getBackendAdapter } from '$lib/services/backend-adapter';
 import type { BackendAdapter } from '$lib/services/backend-adapter';
@@ -70,7 +71,7 @@ async function createNodeHierarchy(
   return { parentId, childIds };
 }
 
-describe.sequential('Section 7: Database Persistence Tests', () => {
+describe.skipIf(!shouldUseDatabase()).sequential('Section 7: Database Persistence Tests', () => {
   let dbPath: string;
   let backend: BackendAdapter;
 

--- a/packages/desktop-app/src/tests/services/event-emission.test.ts
+++ b/packages/desktop-app/src/tests/services/event-emission.test.ts
@@ -33,6 +33,7 @@ import {
   cleanDatabase,
   waitForDatabaseWrites
 } from '../utils/test-database';
+import { shouldUseDatabase } from '../utils/should-use-database';
 import { TestNodeBuilder } from '../utils/test-node-builder';
 import { getBackendAdapter } from '$lib/services/backend-adapter';
 import type { BackendAdapter } from '$lib/services/backend-adapter';
@@ -45,7 +46,7 @@ import type {
 } from '$lib/services/event-types';
 import { sharedNodeStore } from '$lib/services/shared-node-store';
 
-describe.sequential('Section 8: Event System Tests', () => {
+describe.skipIf(!shouldUseDatabase()).sequential('Section 8: Event System Tests', () => {
   let dbPath: string;
   let backend: BackendAdapter;
 

--- a/packages/desktop-app/src/tests/services/node-ordering.test.ts
+++ b/packages/desktop-app/src/tests/services/node-ordering.test.ts
@@ -15,13 +15,14 @@ import {
   cleanDatabase,
   waitForDatabaseWrites
 } from '../utils/test-database';
+import { shouldUseDatabase } from '../utils/should-use-database';
 import { TestNodeBuilder } from '../utils/test-node-builder';
 import { getBackendAdapter } from '$lib/services/backend-adapter';
 import type { BackendAdapter } from '$lib/services/backend-adapter';
 import type { Node } from '$lib/types';
 import { sharedNodeStore } from '$lib/services/shared-node-store';
 
-describe.sequential('Section 6: Node Ordering Tests', () => {
+describe.skipIf(!shouldUseDatabase()).sequential('Section 6: Node Ordering Tests', () => {
   let dbPath: string;
   let backend: BackendAdapter;
 

--- a/packages/desktop-app/vitest.config.ts
+++ b/packages/desktop-app/vitest.config.ts
@@ -164,6 +164,7 @@ export default defineConfig({
 
   test: {
     include: ['src/tests/**/*.{test,spec}.{js,ts}'],
+    exclude: ['src/tests/browser/**'], // Browser tests run separately with test:browser
     environment: 'happy-dom', // Fast, modern DOM for Bun compatibility
     globals: true,
     setupFiles: ['src/tests/setup-svelte-mocks.ts', 'src/tests/setup.ts'],

